test/Microsoft.AspNetCore.Blazor.E2ETest/Tests/ComponentRenderingTest.cs

```diff
@@ -156,6 +156,47 @@ public void CanRenderChildComponents()
             Assert.Equal("somevalue", styledElement.GetAttribute("customattribute"));
         }
 
+        // Verifies we can render HTML content as a single block
+        [Fact]
+        public void CanRenderChildContent_StaticHtmlBlock()
+        {
+            var appElement = MountTestComponent<HtmlBlockChildContent>();
+            Assert.Equal("<p>Some-Static-Text</p>",
+                appElement.FindElement(By.CssSelector("#foo")).GetAttribute("innerHTML"));
+        }
+
+        // Verifies we can rewite more complex HTML content into blocks
+        [Fact]
+        public void CanRenderChildContent_MixedHtmlBlock()
+        {
+            var appElement = MountTestComponent<HtmlMixedChildContent>();
+            Assert.Equal(
+@"
+<!--!-->
+<!--!--><div><p>Some-Static-Text</p></div>
+<!--!--><div id=""bar""><span>More-Static-Text</span></div>
+<div id=""baz""><span>Some-Dynamic-Text</span><!--!--><span>But this is static</span></div>
+
+",
+                appElement.FindElement(By.CssSelector("#foo")).GetAttribute("innerHTML"));
```

Comment:
It's great to have this extra E2E coverage.

It's a bit odd to see the `<!--!-->` details show up in the test though (and below). Elsewhere we've made a series of assertions about the DOM contents rather than snapshotting its HTML representation. It's not like it's a major issue, but the `<!--!-->` markers are a separate implementation detail and could change independently of the child content feature.

Reply:
A further rationale for preferring not to rely on the per-character HTML representation is that if we later want to enable cross-browser tests, we'll find that different browsers pick different (but also valid) HTML representations. For example, having different whitespace around linebreaks, different attribute ordering, or differing rules about when to entity-encode a given character.